compiler/next/include/chpl/queries/stringify-functions.h

```diff
@@ -0,0 +1,298 @@
+/*
+ * Copyright 2021 Hewlett Packard Enterprise Development LP
+ * Other additional copyright holders may be indicated within.
+ *
+ * The entirety of this work is licensed under the Apache License,
+ * Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License.
+ *
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/*
+  This file implements the generic chpl::stringify
+  as well as specializations for some common types.
+
+  stringify can be used to get a string representation of an object
+  for debugging purposes or whenever a string representation of an
+  object is needed
+ */
+
+#ifndef CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+#define CHPL_QUERIES_STRINGIFY_FUNCTIONS_H
+
+#include <cassert>
+#include <cstring>
+#include <functional>
+#include <string>
+#include <unordered_map>
+#include <vector>
+#include "chpl/util/memory.h"
+#include <set>
+#include <sstream>
+#include <algorithm>
+
+
+namespace chpl {
+class Context;
+
+/*
+ * TODO: Implement these definitions
+ * chpl::StringifyKind describes different ways to format the stringification
+ * of an object.
+ * DEBUG_SUMMARY - summary debug output, hash of query,
+                    type information, id/name, ast tag
+ * DEBUG_DETAIL - summary data plus all fields and values, etc.
+ * CHPL_SYNTAX - the chapel syntax necessary to generate the object,
+ *                or DEBUG_DETAIL if unable
+ */
+
+enum StringifyKind {
+  DEBUG_SUMMARY,
+  DEBUG_DETAIL,
+  CHPL_SYNTAX
+};
+
+
+
+// define the generic stringify template which will cause a compilation error
+//  if used. Query argument and return types need to have a specialization of
+//  stringify
+template<typename T> struct stringify {
+  void operator()(std::ostream &stringOut,
+                  StringifyKind stringKind,
+                  const T& stringMe) const = 0;
+};
+
+// define stringify for pointers to call stringify on a reference to avoid
+//  duplication for types that appear both as references and as pointers
+template<typename T> struct stringify<const T*> {
+  void operator()(std::ostream &stringOut,
+                  StringifyKind stringKind,
+                  const T* stringMe) const {
+    stringify<T> stringifier;
+    stringifier(stringOut, stringKind, *stringMe);
+  }
+};
+
+// stringify a vector by stringifying each element; uses [a, b, c] formatting
+template<typename T>
+static inline void defaultStringifyVec(std::ostream &stringOut,
```

Comment:
Our style has primarily been with `&` on the type like `std::ostream& stringOut`. Also the name `stringOut` is a bit misleading because its not a string. Maybe `os`?